### ISSUE TITLE
[Feature Request] Partial object instantiation does not fail with missings

### ISSUE THREAD

**Issue by jzazo** [facebookresearch/hydra#2165]
# 🚀 Feature Request

If a config group has a missing value, it cannot be partially instantiated. I would like that instead of throwing an exception, a partial object is returned where that missing value is a required argument of the partial object.

## Motivation

**Is your feature request related to a problem? Please describe.**
I want to configure a generic hyperparameter search system. At an early stage, hydra instantiates a pipeline but some parameters are missing. During hyperparameter search, the engine proposes new parameters and I want to use them to create full objects to the partially instantiated ones, which work as factories. I would like missing hyperparameters be part of the full object creation. I would like the hyperparmeter search step not know about hydra, which belongs to an outer layer of the pipeline execution.

## Pitch

**Describe the solution you'd like**
```python
#!/usr/bin/env python3
from dataclasses import dataclass
from typing import Any

import hydra
from hydra.core.config_store import ConfigStore
from omegaconf import MISSING, OmegaConf

@dataclass
class PartialFoo:
    _target_: str = "myapp.Foo"
    _partial_: bool = True
    foo: bool = MISSING

@dataclass
class Config:
    partial_object: Any = PartialFoo()

cs = ConfigStore()
cs.store(name="config", node=Config)

@hydra.main(config_path=None, config_name="config")
def my_app(cfg: Config) -> None:
    print(OmegaConf.to_yaml(cfg, resolve=False))
    print(hydra.utils.instantiate(cfg))

if __name__ == "__main__":
    my_app()
```
with `myapp/__init__.py` having class `Foo`:
```
class Foo:
    foo: bool
    boo: str
```
If I comment line 13, i.e., `#    foo: bool = MISSING`, the previous code works.

**Are you willing to open a pull request?**
Yes, with guidance on how to proceed.

**Comment by jieru-hu** [comment]
Thanks for reporting @jzazo , I will take a look at this.

**Comment by jieru-hu** [comment]
@jzazo this is now on the main branch, feel free to give it a try and let me know. (btw you will need java to install Hydra from a branch)

**Comment by Jasha10** [comment]
Tbh I feel a bit concerned about API consistency with this.
Most other times that you touch a MISSING value, a MissingMandatoryValue gets raised (including with attribute/item access as well as non-partial instantiation and OmegaConf.to_object).
It feels weird that MISSING should be silently dropped only if _partial_ is True...

**Comment by jzazo** [comment]
It feels natural that partial instantiation does not fail on `MISSING`s. On the other hand, failing in all cases might be a harsh decision and not flexible enough.

A thought about consistency is that `__get_item__` and `OmegaConf.to_object` are Omegaconf's API responsibilities, whereas `instantiate` and `_partial_` is hydra's API. It makes sense to me to choose to raise error or not depending on the task that hydra supports.

**Comment by Jasha10** [comment]
That's a fair point about the separation between OmegaConf and Hydra.

I guess my concern comes down to `MISSING` being used for two different purposes:
1. MISSING can mean that a given field MUST be supplied.
2. MISSING can mean that a given field MAY be supplied.

For normal instantiation (without `_partial_`), MISSING values give behavior (1), i.e. instantiation will fail fast if there is a missing value. With `_partial_=True` MISSING values now give behavior (2).

**Comment by jieru-hu** [comment]
thanks for the discussions @jzazo @Jasha10 
I'm leaning towards keep the MISSING behavior the same for the non-partial instantiation. The behavior has been consistent since it was introduced and we haven't heard much complaints. _partial_ being a newly introduced feature has a bit more flexibility here.

I thought about this inconsistency while working on the issue. I felt that allowing MISSING in partial seems to be more user friendly which is why i added this.

@jzaro what is the impact on ur use case if we do not allow `foo: bool = MISSING` in a partial object?  Do you mind to share more context here?

**Comment by jzazo** [comment]
I am working on a distributed pipeline for ML in which I want to have step factories and instantiate them fully after certain parameters become available. For example, I may not now the input dimension of a neural network until I have preprocessed the input pipeline, or I may not know the number of output labels until I have selected which labels I want to predict over. What I mean is that I have a lot of runtime parameters that I may or may not know when I instantiate the pipeline. I would like those parameters to belong to the schema, because that's what the object needs to be created (and the schema is static), but I am happy to get a partial object and pass that object along depending on the pipeline context. Furthermore, if we have a meta step like hyperparameter optimization, those partial objects work as factories themselves, simplifying the new pipeline generation.

I find that instantiating objects and partials with missings is a consistent behavior, and a useful one because it adds flexibility. I would lean to not error unless extremely necessary, which I think in this case is not, as it can be delegated to the object itself.

**Comment by Jasha10** [comment]
It sounds like what would really be nice is if the partial object would fail when it is called without the required params:
```python
partial_obj = instantiate(cfg, _partial_=True)
partial_obj()  # raises MissingMandatoryValue: required argument not supplied
```

**Comment by jzaro** [comment]
Hi everyone,
You tagged the wrong person. Please untag me. Thanks.

**Comment by Jasha10** [comment]
@jzazo would something like this work for your use-case?
```python
@hydra.main(config_path=None, config_name="config")
def my_app(cfg: Config) -> None:
    from omegaconf import open_dict

    missings = [
        key
        for key in cfg.partial_object
        if OmegaConf.is_missing(cfg.partial_object, key)
    ]

    for key in missings:
        with open_dict(cfg.partial_object):
            del cfg.partial_object[key]

    print(hydra.utils.instantiate(cfg))  # prints {'partial_object': functools.partial(<class 'myapp.Foo'>)}
```

**Comment by jzazo** [comment]
Yes, I think it would work, thanks for the proposal. However, my opinion is that missing should mean "may be supplied" during instantiation, rather than "must be supplied" (otherwise default values in function signatures are irrelevant, or cannot belong to a schema because they would become required). Thanks for the support on this issue btw.
